torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,231 @@
+                    0,
+                    out_features - orig_out_features,
+                ),
+            )
+            pad_shifted_zero_point = torch.nn.functional.pad(
+                shifted_zero_point,
+                (
+                    0,
+                    new_num_groups - orig_num_groups,
+                    0,
+                    out_features - orig_out_features,
+                ),
+            )
+            return to_affine_quantized_static(
+                input_float=input_float,
+                scale=pad_scale.to(torch.bfloat16),
+                zero_point=pad_shifted_zero_point.to(torch.bfloat16),
+                block_size=block_size,
+                target_dtype=torch.int32,
+                quant_min=quant_min,
+                quant_max=quant_max,
+                zero_point_domain=ZeroPointDomain.FLOAT,
+                layout_type=TensorCoreTiledLayoutType(inner_k_tiles=inner_k_tiles),
+            )
+
+        # TODO(Yi): better way to select the weight quantization function
+        if auto_round_config.bits == 4:
+            weight_func = to_int4_tinygemm_weight
+        else:
+            weight_func = to_uintx_weight
+
+        observed_linear.weight = torch.nn.Parameter(
+            weight_func(observed_linear.weight), requires_grad=False
+        )
+        del observed_linear.scale
+        del observed_linear.zp
+        return observed_linear
+
+    def _is_observed_linear(mod: torch.nn.Module, fqn: str):
+        return hasattr(mod, "scale")
+
+    qmodel = ao_quant.quant_api._replace_with_custom_fn_if_matches_filter(
+        qdq_model, apply_static_quant, _is_observed_linear
+    )
+    return qmodel
+
+
+@ar_utils.dump_elapsed_time()
+@torch.no_grad()
+def apply_auto_round(block, grouped_args, spec, block_outputs):
+    # Call the auto-round to execute the optimization process
+    import auto_round
+
+    ar_utils.see_memory_usage("Before apply auto-round")
+
+    global auto_round_config
+
+    # Start the training process to update the v, alpha and betta.
+    rounder = auto_round.AutoRound(
+        model=block,
+        tokenizer=None,
+        sym=auto_round_config.sym,  # Both True and False are OK
+        bits=auto_round_config.bits,
+        iters=auto_round_config.iters,
+        use_quant_input=False,  # disable it for now
+        amp=auto_round_config.amp,
+        low_gpu_mem_usage=False,
+        model_dtype=next(block.parameters()).dtype,
+    )
+
+    @torch.no_grad()
+    def _unflatten_grouped_args(grouped_args, spec):
+        inputs = []
+        for i, inp in enumerate(grouped_args):
+            cur_args, cur_kwargs = tree_unflatten(inp, spec)
+            inputs.append((cur_args, cur_kwargs))
+        return inputs
+
+    block_inputs = _unflatten_grouped_args(grouped_args, spec)
+    with torch.enable_grad():
+        rounder.quant_block_v2_(
+            block, inputs=block_inputs, outputs=block_outputs, device="cuda"
```

Comment:
btw, do we have to feed the `block_outputs` to the next layer?

Reply:
It's an option that I haven't modified yet. However, some of our preliminary experiments have shown that using or not using the output of the optimized module can impact accuracy.